File: content/examples/other/leetcode/problems/minimum-window-substring/example-2/solution.py
def min_window(s, t):
    if not t:
        return ""
    need = {}
    for ch in t:
        need[ch] = need.get(ch, 0) + 1

    window = {}
    have = 0
    need_kinds = len(need)
    best_len = float("inf")
    best_left = 0
    left = 0

    for right, ch in enumerate(s):
        window[ch] = window.get(ch, 0) + 1
        if ch in need and window[ch] == need[ch]:
            have += 1

        while have == need_kinds:
            length = right - left + 1
            if length < best_len:
                best_len = length
                best_left = left
            drop = s[left]
            window[drop] -= 1
            if drop in need and window[drop] < need[drop]:
                have -= 1
            left += 1

    return "" if best_len == float("inf") else s[best_left:best_left + best_len]


if __name__ == "__main__":
    print(min_window("ADOBECODEBANC", "ABC"))

File: content/examples/other/leetcode/problems/minimum-window-substring/example-2/solution.js
function minWindow(s, t) {
  if (t.length === 0) return "";
  const need = new Map();
  for (const ch of t) need.set(ch, (need.get(ch) || 0) + 1);

  const window = new Map();
  let have = 0;
  const needKinds = need.size;

  let bestLen = Infinity;
  let bestLeft = 0;
  let left = 0;

  for (let right = 0; right < s.length; right += 1) {
    const ch = s[right];
    window.set(ch, (window.get(ch) || 0) + 1);
    if (need.has(ch) && window.get(ch) === need.get(ch)) have += 1;

    while (have === needKinds) {
      const len = right - left + 1;
      if (len < bestLen) {
        bestLen = len;
        bestLeft = left;
      }
      const drop = s[left];
      window.set(drop, window.get(drop) - 1);
      if (need.has(drop) && window.get(drop) < need.get(drop)) have -= 1;
      left += 1;
    }
  }

  return bestLen === Infinity ? "" : s.slice(bestLeft, bestLeft + bestLen);
}

if (require.main === module) {
  console.log(minWindow("ADOBECODEBANC", "ABC"));
}

module.exports = { minWindow };
